drivers/flash/flash_npcx_fiu_nor.c

```diff
@@ -0,0 +1,610 @@
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static int flash_npcx_nor_read(const struct device *dev, off_t addr,
+				 void *data, size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	struct flash_npcx_nor_data *dev_data = dev->data;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || ((size + addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	/* Lock/Unlock SPI bus also for DRA mode */
+	qspi_npcx_fiu_mutex_lock_configure(config->qspi_bus, &config->qspi_cfg,
+					   dev_data->operation);
+
+	/* Trigger Direct Read Access (DRA) via reading memory mapped-address */
+	memcpy(data, (void *)(config->mapped_addr + addr), size);
+
+	qspi_npcx_fiu_mutex_unlock(config->qspi_bus);
+
+	return 0;
+}
+
+static int flash_npcx_nor_erase(const struct device *dev, off_t addr,
+				  size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	int ret = 0;
+	uint8_t opcode;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || (((uint64_t)size + (uint64_t)addr) > config->flash_size)) {
+		LOG_ERR("Addr %ld, size %d are out of range", addr, size);
+		return -EINVAL;
+	}
+
+	/* address must be sector-aligned */
+	if (!SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
+		LOG_ERR("Addr %ld is not sector-aligned", addr);
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sectors */
+	if ((size % config->min_erase_size) != 0) {
+		LOG_ERR("Size %d is not a multiple of sectors", size);
+		return -EINVAL;
+	}
+
+	/* Select erase opcode by size */
+	if (size == config->flash_size) {
+		opcode = SPI_NOR_CMD_CE;
+	} else if (config->min_erase_size == KB(4)) {
+		opcode = SPI_NOR_CMD_SE;
+	} else if (config->min_erase_size == KB(64)) {
+		opcode = SPI_NOR_CMD_BE;
+	} else {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
+
+		if (opcode == SPI_NOR_CMD_CE) {
+			/* chip erase */
+			flash_npcx_uma_cmd_only(dev, opcode);
+			size -= config->flash_size;
+		} else {
+			flash_npcx_uma_cmd_by_addr(dev, opcode, addr);
+			addr += config->min_erase_size;
+			size -= config->min_erase_size;
+		}
+
+		ret = flash_npcx_nor_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static int flash_npcx_nor_write(const struct device *dev, off_t addr,
+				  const void *data, size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	uint8_t *tx_buf = (uint8_t *)data;
+	int ret = 0;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || (((uint64_t)size + (uint64_t)addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		size_t sz_write = size;
+
+		/* Don't write more than a page. */
+		if (sz_write >= SPI_NOR_PAGE_SIZE) {
+			sz_write = SPI_NOR_PAGE_SIZE;
+		}
+
+		/* Don't write across a page boundary */
+		if (((addr + sz_write - 1U) / SPI_NOR_PAGE_SIZE)
+							!= (addr / SPI_NOR_PAGE_SIZE)) {
+			sz_write = SPI_NOR_PAGE_SIZE - (addr % SPI_NOR_PAGE_SIZE);
+		}
```

Comment:
Because your MCU does not have source alignment you can just make this:
```
size_t sz_write;

/* This is max that the first, or any, write can take, which is SP_NOR_PAGE_SIZE */
if (size >= SPI_NOR_PAGE_SIZE) {
		sz_write = SPI_NOR_PAGE_SIZE;
}

/* Correct the first write to not go through page boundary */
size_t overflow = (addr + sz_write) & (SPI_NOR_PAGE_SIZE - 1);
sz_write -= overflow;
/* Now sz_write is reminder needed for addr to align to page boundary */

/* This is still valid test, as nothing has been written yet */
while (size > 0) {
    /* Write something */
    flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
    ret = flash_npcx_uma_write_by_addr(dev, SPI_NOR_CMD_PP,  tx_buf, sz_write, addr);
    if (ret != 0) {
	break;
    }
    ret = flash_npcx_nor_wait_until_ready(dev);
    if (ret != 0) {
        break;
    }
    size -= sz_write;
    addr += sz_write;
    tx_buf += sz_write;
    /* On the first run the sz_write was set to max bytes that would not cross the page boundary,
        but not more than page size and no more than size; for each next run the addr is now aligned to page boundary,
        so try to either write full page or what has left of size */
    if (size > SPI_NOR_PAGE_SIZE) {
        sz_write = SPI_NOR_PAGE_SIZE;
    } else {
         sz_write = size;
    }
}
```
Correct me if I have missed something, and the change is up to you, but I think that this significantly simplifies the loop and reduces branching within it.

Reply:
Thanks for the detailed description. I got your point and will update a new version for it. Regarding `overflow` variable, we might need to check whether the `addr` plus `sz_write` cross the page boundary. Or the size of the first write might be zero.

Reply:
Done.